### ISSUE TITLE
Write a pagination API for GitHub GraphQL queries

### ISSUE THREAD

**Issue by wchargin** [sourcecred/sourcecred#117]
When we request data from the GitHub API, we are limited to a maximum of
100 items per connection at a time: 100 issues per repository, 100
comments per issue, etc. When we want to fetch _all_ data for a
repository, we’ll need to take advantage of GitHub’s pagination API to
fetch subsequent pages. Currently, our code assumes (and asserts) that
all data always fits on a single page, but clearly this is not a
scalable long-term solution: SourceCred itself will soon have more than
100 pull requests, for instance.

Eventually, we would like to implement an “auto-pagination” function
that takes an arbitrary query against the GitHub API, executes it, and
keeps fetching subsequent pages until all data has been exhausted. I’ve
been working on this for a while (order of a week), and it is
fascinatingly difficult. Without going into too much detail: for a fully
general solution, we need to map each field in the response to the field
in the query and the associated GraphQL type, which gets complicated
when dealing with fragments, and further complicated when fragments have
overlapping support. Our existing GitHub query falls into this last
category, so the functionality, at least, is required.

The GraphQL spec is of excellent quality, and is very clear about the
semantics of what happens in the query. This makes it easy enough to
formulate a top-level API and contracts for the functions that we want.
But the implementation of these functions should clearly be decomposed,
and exactly how to decompose them—what types to use, where to cut the
recursion, which base cases to pick—is surprisingly challenging. I’m
confident that I could implement it, given enough time, but it take a
lot of careful thought. (I’d currently estimate one programmer-week’s
worth of design and implementation work, non-parallelizable.)

As our tolerance for ~~hacky~~ ad hoc solutions increases, the problem
becomes easier. In the short term, we should consider implementing a
pagination function that works just for our particular GitHub query.
This removes the greatest difficulty, type tracking across fragments,
because we can design the code around a particular query structure. This
will unblock us indefinitely, at the cost of increasing the friction
involved in changing the GitHub query. (Estimated work: one or two days,
non-parallelizable, including good test coverage; maybe one extra day if
we want really rigorous tests.)

In the long term, it would be nice to implement a more general-purpose,
but still GitHub-specific, pagination API. A good metric for generality
is that it should be releasable as a standalone npm package, with
dependencies only on the structured query language. We can keep this in
mind as it becomes more cumbersome to write queries with manual
pagination—and also as just a “nice-to-have” on the calendar.

Absent objections, I’ll start implementing this; I have the most context
on the existing query system and on the aforementioned pagination
efforts.

cc @decentralion

**Comment by wchargin** [comment]
I’ve now merged PRs #95, #96, #97, #98, #99, and #100, which
collectively implement the pagination API for our particular GitHub
query (as you can tell from the commit spam above!). Here are some
things that I learned along the way and TODOs for the future.

## General foundations

The process of implementing this specific case was valuable. It is
_easier_ to see how to generalize to the general case, and this could
plausibly be done incrementally while maintaining test coverage. It’s
still not trivial: for one thing, the schema is currently hard-coded in
the type of the `enclosingNodeType` attribute of the `Continuation`
type, and in the `continuationsForX` functions.

## Fold direction for continuation accumulation

Suppose that we have a sequence of three queries: the first an original
query, and the next two fetching more pages for some subfields. We might
receive the following results:
```js
const r1 = {
  repository: {
    issues: {
      pageInfo: {hasMorePages: true},
      nodes: [/* issue #1 */],
    },
  },
};
const r2 = {
  _n7: {
    issues: {
      pageInfo: {hasMorePages: true},
      nodes: [/* issue #2 */],
    },
  },
};
const r3 = {
  _n3: {
    issues: {
      pageInfo: {hasMorePages: false},
      nodes: [/* issue #3 */],
    },
  },
};
```

We need to combine these. We have two options: we could combine `r1` and
`r2` to form
```js
const r12 = {
  repository: {
    issues: {
      pageInfo: {hasMorePages: true},
      nodes: [/* issue #1 */, /* issue #2 */],
    },
  },
};
```
and then merge `r3` into this to form our final result, or we could
start by combining `r2` and `r3` thus:
```js
const r23 = {
  _n7: {
    issues: {
      pageInfo: {hasMorePages: true},
      nodes: [/* issue #2 */, /* issue #3 */],
    },
  },
};
```

This can roughly be seen as varying the associativity of the merge
operation: the former option has left-associative merge, and the latter
operator right-associative merge.

The left-associative merge has the advantage that at each step the
result is a sensible response to the original query that becomes
incrementally more complete. This makes it easier to gradually display
more information in a UI, or to pick up where we left off if we are
interrupted. The right-associative merge is more computationally
efficient, because we are always operating on small structures until the
very end, instead of operating on a megastructure at each step.

Which strategy is used is an implementation detail, so there’s nothing
pressing here, but we should keep an eye on it. Currently, we use a
variant of the left-associative merge. In fact, our process is to
generate `r12` and analyze _it_ for continuations to form the query that
yields `r3`, instead of getting these directly from `r2`. This is
especially computationally wasteful, so improving this (while keeping
the merge left-associative) should be our first step if client-side
performance becomes an issue.

## Typing for queries and results

This is an interesting one. In [a comment][comment-flow-types] on #76,
@decentralion proposed that we generate Flow types for the _results_ of
GraphQL queries. If we had had this infrastructure when I was writing
the specific implementation, then a few mistakes that I made would have
been mitigated. Others, that could be caught in principle, would not:
for instance, at one point I declared something like
```js
const subpath = [...path, "issues", i];
```
where I should have written
```js
const subpath = [...path, "issues", "nodes", i];
```
but it is not possible\* to express the semantics of the logical type of
`path` in a Flow type (e.g., that `result.issues.nodes[i]` should be
defined in the above example).

However, I believe that these benefits would be mostly lost if we moved
to the more general API (still GitHub-specific, but not query-specific).
Also, I think that the marginal value of these types _once the
pagination API is already written_ (and never changed again or
refactored, of course! :-) ) is small. The top-level API’s types are
very simple and do not depend on any particular schema types. Therefore,
my conclusion is that my needle on whether we should add GraphQL result
types hasn’t actually moved much at all as a result of performing this
work.

\* Certainly not possible without tricky tricks. I’ve played around with
it and don’t think that it’s possible at all with these value types.
Maybe some contortions would be able to force it, but I suspect that it
would not be worth the trouble.

[comment-flow-types]: https://github.com/sourcecred/sourcecred/issues/76#issuecomment-373613940

## Relay Cursor Connections Specification

GitHub’s GraphQL API implements the [Relay Cursor Connections
Specification][spec]. It would therefore be feasible to write our more
general pagination API against this spec, so that it could be more
easily applicable to other projects. There’s one catch: GitHub provides
some syntactic sugar that is not required by the spec, but which we use
extensively. In particular:

  - The `nodes` fields on connections are not required; only the `edges`
    fields are. Of course, `nodes { … }` can always be replaced with
    `edges { node { … } }`, but this changes the structure of the output
    by introducing an additional level of nesting.

  - The `PageInfo` type is not required to expose `endCursor` and
    `startCursor`; instead, cursor are provided via `edges { cursor }`.
    Here, `data.pageInfo.endCursor` is approximately equivalent to
    `data.edges[data.edges.length - 1].cursor`, modulo error handling.

This presents the following options.

  - Write the pagination API against an extension of the GraphQL spec
    that also requires these extensions. (Before selecting this, it
    would be a good idea to find other implementors of the Relay spec
    and see if they provide a similar affordance. The [examples] in
    the Relay docs have `endCursor` but not `nodes`, for instance.)

  - Write the pagination API against the vanilla Relay spec, and change
    our data format accordingly. This would cause a bit of bloat in the
    output data, but probably not significantly.

  - Write some kind of hybrid that fetches according to the spec but
    then transforms the data into our currently-expected format. (This
    doesn’t seem like a good idea to me.)

[spec]: https://facebook.github.io/relay/graphql/connections.htm
[examples]: https://facebook.github.io/relay/docs/en/graphql-server-specification.html#schema

## Long-term plan

This whole module is predicated on the assumption that recursively
fetching the whole repository is the right thing to do. This scales well
to small, medium, and even large-ish projects, but not to the outliers:
for instance, torvalds/linux has 750,000 commits. Eventually, we’ll need
to make sure to use online algorithms everywhere. But that’s a task for
another time.

**Comment by wchargin** [comment]
I’m closing this issue for now. When we want to extend the pagination
API, we can create a new issue and link it hence.

**Comment by wchargin** [comment]
(But feel free to continue discussing here!)
